Task: Roleauthorizer_v1.1 upgrade

FHieser opened the pull request:
Based on #729

FHieser committed:
Implement Function Lock

FHieser committed:
Implement Function Lock

FHieser committed:
Test Folder restructured

FHieser committed:
Rename folder mistake

FHieser committed:
adapt rest of references

FHieser committed:
Delete TestFolderRefactoring.md

FHieser committed:
Merge branch 'sc-951-task-roleauthorizer_v11-upgrade' of https://github.com/InverterNetwork/contracts into sc-951-task-roleauthorizer_v11-upgrade

FHieser committed:
Adapt functionality

FHieser committed:
Create Tests

FHieser committed:
Include the rest of the new functions

FHieser committed:
Change public Role to Id 1

FHieser committed:
Deprecate Functions

FHieser committed:
Rename to key to Permission

FHieser committed:
Clean up last functions that were still from old implementation

FHieser committed:
Refactor function postions

FHieser committed:
Test Permissioned Modifier

FHieser committed:
Adapt onlyRole to permissioned in Authorizer

FHieser committed:
Cleanup and missed tests

FHieser committed:
Adapt modules to use permissioned modifier

FHieser committed:
Cleanup comments and todos

FHieser committed:
Update Staking Module

FHieser committed:
rename to t.sol

FHieser committed:
Update Recurring Payments

FHieser committed:
Update Paymentrouter

FHieser committed:
Comment Cleanup

FHieser committed:
Add Event references

FHieser committed:
Update KPIRewarder

FHieser committed:
Update Optimistic Oracle

FHieser committed:
Update Bounties

FHieser committed:
Update PP_Streaming

FHieser committed:
Update TokenVault

FHieser committed:
Update BondingCurveBase

FHieser committed:
Adapt comments

FHieser committed:
Update RedeemingBondingCurveBase

FHieser committed:
Update comments

FHieser committed:
Adapt Event references in tests

FHieser committed:
Update Bancor Redeeming

FHieser committed:
Update BondingSurface Redeeming

FHieser committed:
Update comments

FHieser committed:
Update Bonding Surface Redeeming Restricted

FHieser committed:
Fix Typo

FHieser committed:
Update RoleAuthorizerE2E.t.sol

FHieser committed:
Update MetaTxAndMulticallE2E.t.sol

FHieser committed:
Adapt comments

FHieser committed:
Make the test of the upgraded Authorizer the new standard

FHieser committed:
Update VotingRoleManagerE2E.t.sol

FHieser committed:
Update BountyManagerE2E.t.sol

FHieser committed:
Update BondingCurveFundingManagerE2E.t.sol

FHieser committed:
Update BondingCurveTokenRescueE2E.t.sol

FHieser committed:
Update BondingSurfaceFundingManagerE2E.t.sol

FHieser committed:
Update StakingManagerLifecycle.t.sol

FHieser committed:
Update KPIRewarderLifecycle.t.sol

FHieser committed:
Remove Module__CallerNotAuthorized from VotingRoles

FHieser committed:
Remove Module__CallerNotAuthorized from IModule

FHieser committed:
Update comments

FHieser committed:
Remove Authorization from ModuleManagerBase

FHieser committed:
Adapt Orchestrator to use permissioned functions

FHieser committed:
Update KPIRewarderLifecycle.t.sol

FHieser committed:
Make supportsInterface check mandatory for ModuleTest

FHieser committed:
Update Template

FHieser committed:
Adapt E2E tests to take regular Authorizer

FHieser committed:
Remove PIM_Workflowfactory

FHieser committed:
Remove Restricted Bancor Redeeming

FHieser committed:
Remove TokenGatedRoleAuthorizer

FHieser committed:
Adapt comments

FHieser committed:
Merge branch 'dev' into sc-951-task-roleauthorizer_v11-upgrade

FHieser committed:
Fix Merge with dev

0xNuggan commented on src/modules/fundingManager/bondingCurve/FM_BC_Bancor_Redeeming_VirtualSupply_v1.sol:
```diff
@@ -239,9 +240,13 @@ contract FM_BC_Bancor_Redeeming_VirtualSupply_v1 is
         buyingIsEnabled
     {
-        buyFor(_msgSender(), _depositAmount, _minAmountOut);
+        (uint amountIssued, uint collateralFeeAmount) =
```
I see why you did this. Part of the reason we were routing both sell orders (and buy too) through the other was to save on contract size. So if we run into issues on that front again, we may need to refactor the content into a common private function.

0xNuggan reviewed (commented):
This is only for the [FM updates](https://linear.app/inverter/issue/SC-998/task-review-authorizer-update-fundingmanager-modules). Looks good in general. Although since we will do a major version upgrade, we are missing the removal of all the deprecated functions/variables

0xNuggan reviewed (changes requested):
This is only for the [FM updates](https://linear.app/inverter/issue/SC-998/task-review-authorizer-update-fundingmanager-modules). Looks good in general. Although since we will do a major version upgrade, we are missing the removal of all the deprecated functions/variables

FHieser committed:
Make FM_Oracle and PP_Queue buildable

FHieser committed:
Adapt correct file ending

FHieser committed:
Update OracleFundingManagerAndManualQueueBasedPaymentProcessorE2E.t.sol

FHieser committed:
Test modifier in places

FHieser committed:
Do leftover todos

- Removing leftover Role declarations
- moving Authorization comment to downstream implementation
- cleanup comments

FHieser commented on src/modules/fundingManager/bondingCurve/FM_BC_BondingSurface_Redeeming_Restricted_Repayer_Seizable_v1.sol:
```diff
@@ -431,70 +330,56 @@ contract FM_BC_BondingSurface_Redeeming_Restricted_Repayer_Seizable_v1 is
 
+    /// @inheritdoc IFM_BC_BondingSurface_Redeeming_Restricted_Repayer_Seizable_v1
+    function restrictBuyAndSell() external {
+        revert IModule_v1.Module__FunctionDeprecated();
```
The wording on this functionality was not that good (my bad)
This implemented a role access restriction in the background. The turn off buy and sell functionality is still implemented in the base contract and not affected by this change

FHieser committed:
Adapt based on Pablos comments

FHieser committed:
Remove deprecated functions

FHieser committed:
Fix test

FHieser committed:
Revert "Remove TokenGatedRoleAuthorizer"

This reverts commit b5fa277c016e18d0bf10132fd1304ac8592bccb8.

FHieser committed:
Adapt comments

FHieser committed:
Adapt Token Gated Roles

FHieser committed:
Update TokenGatedRoleAuthorizerE2E.t.sol

FHieser committed:
Update AUT_TokenGated_Roles_v1.sol

FHieser committed:
Adapt tests

FHieser committed:
Turn down the rate of assume problems

0xNuggan commented on src/modules/authorizer/role/interfaces/IAUT_EXT_VotingRoles_v1.sol:
```diff
@@ -44,6 +44,9 @@ interface IAUT_EXT_VotingRoles_v1 {
     //--------------------------------------------------------------------------
     // Errors
 
+    /// @notice Theis function is only callable by this contract.
```
Typo

0xNuggan commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,181 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+    // ========================================================================
+    // Events
+
+    /// @notice Emits when a function permission is added to a role.
```
Kind of nitpicky but I would invert the order of the explanation to avoid confusion. What we are doing in the code is adding a role to a function, not a function selector to a role. This way the order of parameters makes more sense also.

0xNuggan commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,181 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+        address target, bytes4 functionSelector, bytes32 roleId
+    );
+
+    /// @notice Emits when a function permission is removed from a role.
```
Same as above

0xNuggan commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,181 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+    /// @param  selector_ The selector of the function.
+    /// @param  roleId_ The roleId that we want to check.
+    /// @return isPermissioned_ Returns if the roleId is permissioned to call the function.
+    function isPermissioned(address target_, bytes4 selector_, bytes32 roleId_)
```
why not "hasPermission(role, target, selector)"? I'd asssume this function will mostly be called from the perspectiva of a role holder, right?

0xNuggan commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,181 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+    /// @param  selector_ The selector of the function.
+    /// @param  roleId_ The roleId that we want to check.
+    /// @return isPermissioned_ Returns if the roleId is permissioned to call the function.
+    function isPermissioned(address target_, bytes4 selector_, bytes32 roleId_)
```
"isPermissioned"to me suggests a split between public/permissioned. I think it would be clearer to override hasPermission below, and have (roleId, target, selector) as parameters. They both serve the same purpose ("check whether X can call this function"). Or at least rename this function to be more clear that it's checking for a specific role.

FHieser replied:
Yeah I see the point
I renamed it to isRolePermissioned, which should make it a bit more clear

0xNuggan commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,181 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+    ) external returns (bytes32 _newRoleId);
+
+    /// @notice Changes the name of a role.
+    /// @dev    Labels are emitted as events and therefor are not accessible
```
typo

0xNuggan commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,181 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+    /// @dev    Function access controlled by authorizer.
+    /// @dev    The role of the admin has to be created already.
+    /// @dev    The array of targets corresponds with the two dimensional array of selectors.
+    ///         The first position of targets therefor is assigned to the first position of the selector array.
```
Typo

0xNuggan replied:
https://www.grammarly.com/blog/commonly-confused-words/therefore-vs-therefor/ 🤓

0xNuggan commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -35,6 +35,10 @@ import {AccessControlEnumerableUpgradeable} from
  *                          In case of any concerns or findings, please refer to our Security Policy
  *                          at security.inverter.network or email us directly!
  *
+ * @custom:version  v1.1.0
```
I guess this will change now :grin:

FHieser replied:
Yes I didnt want to do this yet, as a change in the major version would mean, that I have to adapt the file name itself, which would prevent you from actually checking the differences between the files themselves.
This is the real final step of the PR and I will do that last

0xNuggan commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -54,50 +58,52 @@ contract AUT_Roles_v1 is
+
+    /// @notice The counter for role IDs.
+    /// @dev	This is used to generate unique role IDs for each role.
+    /// @dev    Starts at 1, which symbolizes two roles: PUBLIC_ROLE and DEFAULT_ADMIN_ROLE,
```
I guess this is pretty much semantics, but wouldn't DEFAULT_ADMIN_ROLE and PUBLIC_ROLE be two roles at the end of the day? So the counter should start at two, and then have an assign->increment flow instead of increment->assign as is now?

FHieser replied:
Maybe the naming of the counter itself is chosen a bit poorly.
It essentially represents the last created Id instead of the exact amount of ids.
Because the DefaultAdmin Role has to start at 0 and the next role would be the already created public role it is set to 1 as the public role is technically the last created role.
The additional purpose is to compare wether given ids are valid, because they have to be either equal or lower to the value to be valid.
Would it help if we named it something like lastCreatedId ?

0xNuggan replied:
yes, something like lastAssignedRoleId() would be more clear. Specially for the public getter function

FHieser replied:
Renamed it to lastAssignedRoleId

0xNuggan commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -135,164 +144,246 @@ contract AUT_Roles_v1 is
+        virtual
+        returns (bool hasPermission_)
+    {
+        // If caller is the admin, they can call any function.
```
Just to double check: did we agree on this in the final spec?

FHieser replied:
Yes
you can look it up in the google doc under "Role Auth V2 Living Documentation"

0xNuggan commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -54,50 +58,52 @@ contract AUT_Roles_v1 is
-            revert Module__Authorizer__AdminRoleCannotBeEmpty();
+    /// @dev     Verifies that the roleId is already existing.
+    /// @param  roleId_ The id of the role.
+    modifier idExisting(bytes32 roleId_) {
```
would rename to "idExists"

0xNuggan commented on test/e2e/logicModule/BountyManagerE2E.t.sol:
```diff
@@ -107,10 +111,65 @@ contract BountyManagerE2E is E2ETest {
-        bountyManager.grantModuleRole(
-            bountyManager.BOUNTY_ISSUER_ROLE(), address(this)
+        // =========
+        // Setting up Roles
```
I think we discussed this, but just to bring it up in writing: I think something like this should be part of the base Module in the form of an abstract function. So on initialization or right afterwards it gets executed to at least configure the role setup (register+setAdmin), even if the roles have no users yet.

FHieser commented on src/modules/base/IModule_v1.sol:
```diff
@@ -23,34 +23,11 @@ interface IModule_v1 {
-
-    /// @notice Function is only callable by the {Orchestrator_v1}.
-    error Module__OnlyCallableByOrchestrator();
+    error Module__NotPermissioned();
```
Renamed it to Module__CallerNotPermissioned

0xNuggan replied:
*chef's kiss*

FHieser committed:
Addressed Pablos comments

FHieser committed:
Fix test

It was the ||

FHieser committed:
Adapt the function positioning

FHieser committed:
Adapt natspec

FHieser committed:
Fix typo

Damn copy paste

FHieser committed:
Fix test for real

marvinkruse commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -135,164 +340,246 @@ contract AUT_Roles_v1 is
     /// @inheritdoc IAuthorizer_v1
-    function generateRoleId(address module, bytes32 role)
+    function getPermissions(address target_, bytes4 selector_)
         public
```
I can't find any place where this is used internally, so can be external? (May also be true for other functions, but maybe i also oversaw something where its used)

marvinkruse commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -135,164 +340,246 @@ contract AUT_Roles_v1 is
+                return true;
+            }
+        }
+        // Caller does not have any of the roles, so they cannot call the function.
```
```suggestion
        // Caller does not have any of the roles, so they cannot call the
        // function.
```

marvinkruse commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -135,164 +340,246 @@ contract AUT_Roles_v1 is
 
     /// @inheritdoc IAuthorizer_v1
-    function grantGlobalRole(bytes32 role, address target)
+    function burnAdminFromRole(bytes32 roleId_)
```
To me it feels this should be called `burnAdminOfRole` rather than `fromRole`. Maybe a bit too pedantic though, but you are the admin `of` something, not `from` something. Idk if that's just me though.

FHieser replied:
burnRoleAdmin equals burnAdminOfRole for me,
So my reasoning still stands here

marvinkruse replied:
I still find the naming confusing, but I get where you are coming from. While I focus on the relationship being severed (remove admin OF role) you focus on the action of removal (remove admin FROM role). But I still find "burnAdminOfRole" too confusing, as it can also be understood as "burning the admin of a role" and not "burning the admin role of a role" which it technically is.
I'd also argue in favor of the idea from @0xNuggan to call it "burnRoleAdmin" (which is NOT the same as burnAdminOfRole, declaring them to be the same linguistically makes no sense to me, as admin and an admin role are not the same) - as that is 100% clear to me.

FHieser replied:
Okay then I will adapt it :+1:

marvinkruse commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -135,164 +340,246 @@ contract AUT_Roles_v1 is
 
-    //--------------------------------------------------------------------------
-    // Overloaded and overridden functions
+        // Run through all target and selector combinations and add permission to role id
```
```suggestion
        // Run through all target and selector combinations and add permission
        // to role id.
```

marvinkruse reviewed (changes requested):
Couldn't find any logical issues really, good work! Looks really good to me, just some minor things to be annoying.

FHieser committed:
Update comments

0xNuggan commented on src/modules/authorizer/role/AUT_TokenGated_Roles_v1.sol:
```diff
@@ -95,201 +105,206 @@ contract AUT_TokenGated_Roles_v1 is IAUT_TokenGated_Roles_v1, AUT_Roles_v1 {
+        idExists(roleId_)
+        onlyEmptyRole(roleId_)
+        notPublicRole(roleId_)
+    // @todo Do we prevent default admin here?
```
Yeah, I also think it can stay like that

FHieser committed:
Adapt natspec

FHieser committed:
Update Natspec for Token Gated Roles

FHieser commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,182 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+        string memory roleName_,
+        bytes32 respectiveAdminRole_,
+        address[] memory initialMembers_
+    ) external returns (bytes32 _newRoleId);
```
```suggestion
    ) external returns (bytes32 newRoleId_);
```

FHieser committed:
Apply suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

FHieser committed:
Apply suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

FHieser committed:
Rename RoleIdCounter to LastAssignedRoleId

FHieser committed:
Merge branch 'sc-951-task-roleauthorizer_v11-upgrade' of https://github.com/InverterNetwork/contracts into sc-951-task-roleauthorizer_v11-upgrade

FHieser commented on src/modules/authorizer/extensions/AUT_EXT_VotingRoles_v1.sol:
```diff
@@ -45,11 +45,9 @@ contract AUT_EXT_VotingRoles_v1 is IAUT_EXT_VotingRoles_v1, Module_v1 {
 
     /// @dev	Reverts if caller is not the module itself.
-    modifier onlySelf() {
+    modifier onlyCallableByMotion() {
```
I see your point here, that technically this is not restricted to motions, but to be only callable by the module itself, but the intention of the modifier is that it is actually only callable by the motions of the module.
Also Pablo already criticized the naming before ([here](https://github.com/InverterNetwork/contracts/pull/730#discussion_r2084104164) ), so im not quite sure how to name this properly.
If you have a better name please let me know :see_no_evil:

0xNuggan replied:
Maybe my comment wasn't clear, apologies. I think that if the modifier controls for "self" to be calling, `onlySelf` is an appropriate name. My nitpick was more with the error name, which could be more explicit in saying "this is reverting because this function can only be called by the contract itself"

FHieser replied:
okay then I will revert it to onlySelf and adapt the error to OnlySelfCallAllowed

FHieser commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -54,50 +59,52 @@ contract AUT_Roles_v1 is
-    modifier onlyModule(address module) {
-        if (!orchestrator().isModule(module)) {
-            revert Module__Authorizer__NotActiveModule(module);
+    modifier idNotDefaultAdmin(bytes32 roleId_) {
```
I personally like to put access or parameter restrictions into modifier, even if they are only used a single time.
Fits better in my personal workflow (how i work).
Is there a reason that I should not do it like that?

FHieser commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -54,50 +59,52 @@ contract AUT_Roles_v1 is
+    // Storage
 
-    //--------------------------------------------------------------------------
+    /// @notice The role that is used as a placeholder for a burned admin role.
```
Yeah this was a leftover when I was not sure which number the Burnadmin role would get.
Just copy pasted the comment for the Public role

FHieser commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -54,50 +59,52 @@ contract AUT_Roles_v1 is
+    bytes32 public constant BURN_ADMIN_ROLE =
+        0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff;
+
+    /// @notice The role that is used as a placeholder for a public role.
```
See comment above

FHieser commented on src/modules/authorizer/role/AUT_Roles_v1.sol:
```diff
@@ -135,164 +145,245 @@ contract AUT_Roles_v1 is
-        _setRoleAdmin(roleId, BURN_ADMIN_ROLE);
+        newRoleId_ = bytes32(++_roleIdCounter);
+
+        emit RoleCreated(newRoleId_, roleName_);
```
For me the moment the counter is increased the role is technically created.
The setting of the admin and the granting of roles where different "functionality blocks".
It felt right to place the emit in the "Creation Block".
Do you want me to change it?

marvinkruse replied:
Yeah, alright, your argument makes sense indeed. I think it's fine then!

FHieser committed:
Address Marvins comments

FHieser committed:
Update Voting Roles variable naming

FHieser committed:
Adapt to inverter standard

FHieser commented on src/modules/authorizer/IAuthorizer_v1.sol:
```diff
@@ -21,99 +50,182 @@ interface IAuthorizer_v1 is IAccessControlEnumerable {
+    // ------------------------------------------------------------------------
+    // Getter -  Authorization
+
+    /// @notice Returns the permissions of the given function in the target contract.
```
Adapted Aut_Roles TokenGated and VotingRoles.

FHieser committed:
Adapt OnlyCallableByMotion modifier and error

marvinkruse reviewed (changes requested):
Answered where I could find open questions and added a few tiny things more. 
Let me know if I missed anything you tagged me in or needed feedback on! @FHieser

FHieser commented on script/workflowDeploymentAndSetupScripts/DeployAndSetupNavBasedPimWorkflow.s.sol:
```diff
@@ -1,4 +1,6 @@
-// SPDX-License-Identifier: UNLICENSED
+//@todo how to adapt this?
```
Yeah forgot to write you here.
The Deploymentscript itslef relies heavily on the old Authorizer for the predetermined Roles Setup.
I initially wasnt sure how to proceed on this, but I already talked to Marvin G about this.
As the script is specific to the Piku deployments and technically not needed anymore after the deployments are done,
I would scrap it for now, as we can reference it in the "Before Authorizer Update" Legacy Branch in the planned Repo

FHieser replied:
Do you think that fits here?

FHieser committed:
Apply suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

FHieser committed:
Adapt comments

FHieser committed:
Remove Pim Workflow as it is outdated

FHieser committed:
Adapt comments

marvinkruse commented on test/unit/modules/authorizer/role/AUT_Roles_v1.t.sol:
```diff
@@ -1,1147 +1,1181 @@
+            └── Then: The Admin should be burned
+    */
+
+    function testburnRoleAdmin_OnlyRoleAdmin(
```
```suggestion
    function testBurnRoleAdmin_OnlyRoleAdmin(
```

marvinkruse commented on test/unit/modules/authorizer/role/AUT_Roles_v1.t.sol:
```diff
@@ -1,1147 +1,1181 @@
-                _authorizer.generateRoleId(newModule, ROLE_0), BOB
-            ),
-            false
+    function testburnRoleAdmin_ModifierInPositionChecks() public {
```
```suggestion
    function testBurnRoleAdmin_ModifierInPositionChecks() public {
```

marvinkruse commented on test/unit/modules/authorizer/role/AUT_Roles_v1.t.sol:
```diff
@@ -1,1147 +1,1181 @@
-            _authorizer.generateRoleId(address(_orchestrator), bytes32("0x03"));
-        vm.prank(ALBA);
-
+    function testburnRoleAdmin_idExists(bytes32 roleId_, bytes32 adminRoleId_)
```
```suggestion
    function testBurnRoleAdmin_idExists(bytes32 roleId_, bytes32 adminRoleId_)
```

marvinkruse reviewed (changes requested):
Just some capitalization issue after ctrl+f replacements

FHieser committed:
Apply suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

FHieser committed:
Update Authorizer Comments.md

FHieser committed:
Delete Authorizer Comments.md

FHieser committed:
Merge branch 'sc-951-task-roleauthorizer_v11-upgrade' of https://github.com/InverterNetwork/contracts into sc-951-task-roleauthorizer_v11-upgrade

0xNuggan commented on src/modules/authorizer/role/AUT_TokenGated_Roles_v1.sol:
```diff
@@ -21,304 +21,444 @@ import {AccessControlUpgradeable} from
+ *                    achieve this. Without going into too much detail, this
+ *                    is the main part that is needed to understand the basic
+ *                    mechanism:
+ *                    In the contract the token gating is done by instead of
```
This explanation is a bit confusing to me. Suggestion: "In the contract, token gating is implemented by storing the token address in the role’s members property, instead of directly listing user addresses. This setup allows the contract to check the token balance of a user against a defined threshold when access is requested.
Example: We want to restrict a role to users who hold a certain amount of Token A. Therefore, we configure the role to be token-gated and set a required token amount the user needs to hold as threshold. When we then call 'grantRole' with the address of Token A, the role becomes accessible only to users whose wallet holds at least the specified amount of that token."

FHieser committed:
Adapt documentation of TokenGated_Roles
